File: app/src/main/java/com/github/ykiselev/playground/services/fs/ArchiveFileSystem.java
package com.github.ykiselev.playground.services.fs;

import com.github.ykiselev.assets.ResourceException;
import com.github.ykiselev.common.closeables.Closeables;
import com.github.ykiselev.spi.services.FileSystem;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.channels.FileChannel;
import java.nio.channels.ReadableByteChannel;
import java.nio.file.FileSystems;
import java.nio.file.OpenOption;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.stream.Stream;

/**
 * Wrapper for {@link java.nio.file.FileSystem} created from zip archive
 */
public final class ArchiveFileSystem implements FileSystem {

    private final Logger logger = LoggerFactory.getLogger(getClass());

    private final java.nio.file.FileSystem fs;

    public ArchiveFileSystem(java.nio.file.FileSystem fs) {
        this.fs = Objects.requireNonNull(fs);
    }

    @Override
    public void close() {
        Closeables.close(fs);
    }

    @Override
    public Optional<ReadableByteChannel> open(String resource) throws ResourceException {
        return Optional.ofNullable(resource)
                .map(name -> {
                    try {
                        return FileChannel.open(fs.getPath(resource), StandardOpenOption.READ);
                    } catch (IOException e) {
                        logger.error("Failed to open file!", e);
                        return null;
                    }
                });
    }

    @Override
    public Stream<ReadableByteChannel> openAll(String resource) throws ResourceException {
        return open(resource).stream();
    }

    @Override
    public FileChannel open(String name, OpenOption... options) {
        try {
            return FileChannel.open(fs.getPath(name), options);
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }

    @Override
    public FileSystem mapArchive(String archiveName, boolean create, boolean readOnly) {
        return ArchiveFileSystem.create(fs.getPath(archiveName), create);
    }

    public static FileSystem create(Path path, boolean create) {
        final java.nio.file.FileSystem fs;
        try {
            fs = FileSystems.newFileSystem(path, Map.of("create", Boolean.toString(create)));
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
        return new ArchiveFileSystem(fs);
    }
}

File: app/src/main/kotlin/com/github/ykiselev/playground/services/fs/ArchiveFileSystem.kt
package com.github.ykiselev.playground.services.fs

import com.github.ykiselev.assets.ResourceException
import com.github.ykiselev.common.closeables.Closeables
import org.slf4j.LoggerFactory
import java.io.IOException
import java.net.URI
import java.nio.channels.FileChannel
import java.nio.channels.ReadableByteChannel
import java.nio.file.*
import java.util.*

/**
 * Wrapper for [java.nio.file.FileSystem] created from zip archive
 */
class ArchiveFileSystem(private val fs: FileSystem) : com.github.ykiselev.spi.services.FileSystem {

    private val logger = LoggerFactory.getLogger(javaClass)

    override fun close() {
        Closeables.close(fs)
    }

    @Throws(ResourceException::class)
    override fun open(resource: String): ReadableByteChannel? =
        try {
            FileChannel.open(fs.getPath(resource), StandardOpenOption.READ)
        } catch (e: IOException) {
            logger.error("Failed to open file!", e)
            null
        }


    @Throws(ResourceException::class)
    override fun openAll(resource: String): Sequence<ReadableByteChannel> =
        open(resource)?.let { sequenceOf(it) } ?: emptySequence()

    override fun open(name: String, vararg options: OpenOption): FileChannel =
        FileChannel.open(fs.getPath(name), *options)

    override fun mapArchive(
        archiveName: String,
        create: Boolean,
        readOnly: Boolean
    ): com.github.ykiselev.spi.services.FileSystem =
        create(fs.getPath(archiveName), create)

    companion object {

        @JvmStatic
        fun create(path: Path, create: Boolean): com.github.ykiselev.spi.services.FileSystem =
            path.let {
                Files.createDirectories(it.parent)
                ArchiveFileSystem(
                    FileSystems.newFileSystem(
                        URI.create("jar:file:$path"),
                        mapOf("create" to create.toString())
                    )
                )
            }
    }
}